File: rxui-sample-java/src/main/java/com/artemzin/rxui/sample/java/AuthService.java
package com.artemzin.rxui.sample.java;


import android.support.annotation.NonNull;

import rx.Observable;

interface AuthService {

    class Success {

    }

    class Failure {

    }

    @NonNull
    Observable<Object> signIn(@NonNull String login, @NonNull String password);

    class Impl implements AuthService {

        @NonNull
        @Override
        public Observable<Object> signIn(@NonNull String login, @NonNull String password) {
            return Observable.fromCallable(() -> (login.equals("@artem_zin") && password.equals("rxui") ? new Success() : new Failure()));
        }
    }
}

File: rxui-sample-kotlin/src/main/kotlin/com/artemzin/rxui/sample/kotlin/AuthService.kt
package com.artemzin.rxui.sample.kotlin

import com.artemzin.rxui.sample.kotlin.AuthService.Response.Failure
import com.artemzin.rxui.sample.kotlin.AuthService.Response.Success
import rx.Observable

interface AuthService {

    // Kind of algebraic data type, compiler will check that we handle all cases and so on.
    sealed class Response {
        object Success : Response()
        class Failure(val cause: Throwable) : Response()
    }

    fun signIn(login: String, password: String): Observable<Response>

    class Impl : AuthService {
        override fun signIn(login: String, password: String) = Observable.fromCallable {
            if (login == "@artem_zin" && password == "rxui") Success
            else Failure(IllegalArgumentException("Incorrect credentials"))
        }
    }
}
